nixos/ncps: delete failing secretKeyPath assertion and wait for network-online.target

headblockhead opened the pull request:
The assertion `config.ncps.cache.secresecretKeyPath=/run/secrets/NCPSSecretKey must exist but does not` is:

1. miss-spelled
2. incompatible with the `--target-host` argument to `nixos-rebuild`
2. incompatible with secret management systems like `agenix`

It fails when the machine **evaluating** the module does not have the secret key in its filesystem, even if the machine running the service does have the secret key file.

This assertion pattern isn't seen in other modules that have secret files, for example wireshark.

Therefore, I believe it should be removed. 

fixes: #395214

Also, the ncps service will fail to start if it cannot reach the upstream caches when starting, so waiting for the network to be online ensures it will start successfully when a machine is booting.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

pizzapim commented:
I think it has even more problems, namely:
```
++ (lib.optional (cfg.cache.secretKeyPath != null) "--cache-secret-key-path='%d/secretKey'")
```

`%d` will evaluate to the Systemd unit's [credentials directory](https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html), which is ` /run/credentials/ncps.service`. I think the line should instead say:
```
++ (lib.optional (cfg.cache.secretKeyPath != null) "--cache-secret-key-path='${cfg.cache.secretKeyPath}'")
```

headblockhead commented:
Good spot! This module has a surprising amount of errors.

I'll add your changes to this PR.

I also already have another PR open to fix a different error in this module (#397685), but I'll close that PR and put the changes it made here, making a single PR to fix the whole module at once.

headblockhead committed:
nixos/ncps: wait for network-online, remove failing assertion

- Start after network-online as ncps requires all upstreams to be online
  and reachable when starting
- Remove a bad assertion causing failures at eval time when using
  secrets management systems

headblockhead commented:
Just adding in all the changes now, but having re-read the Nix module I believe the original
```nix
++ (lib.optional (cfg.cache.secretKeyPath != null) "--cache-secret-key-path='%d/secretKey'")
```
is actually correct, as this modules uses [systemd's LoadCredential](https://systemd.io/CREDENTIALS/), which is used is many other nix modules, and was discussed in #102397 as being a valid secrets storing method.

The secret is set [here](https://github.com/NixOS/nixpkgs/blob/af4b8db3de07e90fee483f94429f8c83a60ad751/nixos/modules/services/networking/ncps.nix#L260C1-L262C11).

headblockhead commented:
Tried to rename the branch, turns out GitHub won't let you change the head branch of a PR. :sweat_smile:

headblockhead commented:
@kalbasit

kalbasit commented:
@headblockhead thank you for getting this fixed, much appreciated. The change looks good to me, although, I did not test it myself. Does it work as expected for you?

kalbasit commented:
I ran the nixos test myself and it's working fine.

headblockhead commented:
Thanks for merging!
